### PR TITLE
M1 Firestorm model

### PR DESCRIPTION
**Pull request by tomhepworth**
Contains my best performing config fire (m1_firestorm6.yaml from my documentation on branch `m1-dev`) and a port allocator class that aims to approximate the load balancing of the firestorm core's front-end.

### CODE REVIEW

File: src/include/simeng/pipeline/M1PortAllocator.hh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  /** Allocate the lowest weighted port available for the specified instruction
+   * group. Returns the allocated port, and increases the weight of the port.
+   */
+  uint8_t allocate(const std::vector<uint8_t>& ports) override;
+
+  /** Decrease the weight for the specified port. */
+  void issued(uint8_t port) override;
+
+  /** Decrease the weight for the specified port. */
+  void deallocate(uint8_t port) override;
```
Comment (FinnWilkinson) [inline src/include/simeng/pipeline/M1PortAllocator.hh:33]: Do these functions do the same thing? Or just duplicated comments?

File: src/lib/pipeline/M1PortAllocator.cc [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  return bestPort;
+}
+
+void M1PortAllocator::issued(uint8_t port) {
+  assert(weights[port] > 0);
+  weights[port]--;
+}
+
+void M1PortAllocator::deallocate(uint8_t port) { issued(port); };
+
```
Comment (FinnWilkinson) [inline src/lib/pipeline/M1PortAllocator.cc:62]: What is the need for both functions if one simply calls the other?

**Review by FinnWilkinson — CHANGES_REQUESTED** [review]
Bit of confusion over `issued` and `deallocate` functions in the M1 port allocator

**Review by jj16791 — APPROVED** [review]

File: src/include/simeng/pipeline/M1PortAllocator.hh (hunk 1 above)
Comment (FinnWilkinson) [reply src/include/simeng/pipeline/M1PortAllocator.hh:33]: IGNORE - Clarified by Jack

File: src/lib/pipeline/M1PortAllocator.cc (hunk 2 above)
Comment (FinnWilkinson) [reply src/lib/pipeline/M1PortAllocator.cc:62]: IGNORE - Clarified by Jack

**Review by FinnWilkinson — APPROVED** [review]

**Pull request merged**